src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -149,13 +157,24 @@ public virtual bool StartSession(ITestSessionEventsHandler eventsHandler, IReque
             stopwatch.Elapsed.TotalSeconds);
 
         // Dispose of all proxies if even one of them failed during setup.
+        //
+        // Update: With the introduction of the proxy creation fail disposal policy, we now support
+        // the scenario of individual proxy setup failures. What this means is that we don't mark
+        // the whole session as failed if a single proxy fails, but instead we'll reuse the spinned
+        // off testhosts when possible and create on-demand testhosts for the sources that we failed
+        // to create proxies for.
         if (_proxySetupFailed)
         {
-            requestData?.MetricsCollection.Add(
-                TelemetryDataConstants.TestSessionState,
-                TestSessionState.Error.ToString());
-            DisposeProxies();
-            return false;
+            if (DisposalPolicy == ProxyDisposalOnCreationFailPolicy.DisposeAllOnFailure)
+            {
+                requestData?.MetricsCollection.Add(
+                    TelemetryDataConstants.TestSessionState,
+                    TestSessionState.Error.ToString());
+                DisposeProxies();
+                return false;
+            }
+
+            EqtTrace.Info($"ProxyTestSessionManager.StartSession: At least one proxy setup failed, but failures are tolerated by policy.");
```

Comment:
Can we improve this one with the list of failed proxy/hosts version?

Reply:
We can, but I wanted to avoid doing this since some synchronization is going to be needed when inserting in the list, which will only make the session creation process slower.